File: app/src/test/java/com/beautifulquran/ui/theme/ornament/CoverMonogramTest.kt
package com.beautifulquran.ui.theme.ornament

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import kotlin.math.PI
import kotlin.math.hypot

class CoverMonogramTest {

    private val mark = CoverMonogram

    @Test
    fun `no enclosing circle — every stroke is an open path`() {
        assertTrue(mark.strokes.isNotEmpty())
        for (s in mark.strokes) {
            assertFalse("cover mark must not close into a ring: $s", s.closed)
        }
        assertTrue(mark.dots.isEmpty())
    }

    @Test
    fun `crescent is two offset open arcs, not concentric rings`() {
        val arcs = mark.strokes.filter { it.points.size >= 20 }
        assertEquals("two crescent arcs", 2, arcs.size)
        val c0 = arcCenter(arcs[0])!!
        val c1 = arcCenter(arcs[1])!!
        val offset = hypot(c0.x - c1.x, c0.y - c1.y)
        assertTrue("crescent centres must be offset, not concentric: $offset", offset > 0.02)
        for (arc in arcs) {
            val sweep = arcSweep(arc)
            assertTrue("arc must not be a full circle (sweep=$sweep)", sweep < 2.0 * PI - 0.4)
            assertTrue("arc must cover the left of the moon (sweep=$sweep)", sweep > PI)
        }
    }

    @Test
    fun `geometry stays in the unit box`() {
        for (s in mark.strokes) {
            assertTrue(s.points.size >= 2)
            assertTrue(s.birth >= 0.0 && s.birth + s.span <= 1.0001)
            for (p in s.points) {
                assertTrue("point out of unit box: $p", p.x in -0.001..1.001)
                assertTrue("point out of unit box: $p", p.y in -0.001..1.001)
            }
        }
    }

    @Test
    fun `allah sits in the opening — right of the crescent's inner centre`() {
        val inner = mark.strokes.filter { it.points.size >= 20 }.minBy { it.points.size }
        val c = arcCenter(inner)!!
        val lettering = mark.strokes.filter { it.points.size < 20 }
        assertTrue(lettering.size >= 4)
        val midX = lettering.flatMap { it.points }.map { it.x }.average()
        assertTrue("الله should sit in the crescent's opening (right), midX=$midX c.x=${c.x}", midX > c.x)
    }
}

File: web/src/ui/theme/__tests__/coverMonogram.test.ts
import { describe, expect, it } from 'vitest'
import { coverMonogram } from '../coverMonogram'

describe('coverMonogram', () => {
  it('has no enclosing circle — every stroke is an open path', () => {
    expect(coverMonogram.strokes.length).toBeGreaterThan(0)
    for (const s of coverMonogram.strokes) {
      expect(s.closed).toBe(false)
    }
    expect(coverMonogram.dots).toEqual([])
  })

  it('is two offset open arcs, not concentric rings', () => {
    const arcs = coverMonogram.strokes.filter((s) => s.points.length >= 20)
    expect(arcs).toHaveLength(2)
    const c0 = arcCenter(arcs[0]!)
    const c1 = arcCenter(arcs[1]!)
    const offset = Math.hypot(c0.x - c1.x, c0.y - c1.y)
    expect(offset).toBeGreaterThan(0.02)
    for (const arc of arcs) {
      const sweep = arcSweep(arc)
      expect(sweep).toBeLessThan(2 * Math.PI - 0.4)
      expect(sweep).toBeGreaterThan(Math.PI)
    }
  })

  it('stays in the unit box', () => {
    for (const s of coverMonogram.strokes) {
      expect(s.points.length).toBeGreaterThanOrEqual(2)
      expect(s.birth).toBeGreaterThanOrEqual(0)
      expect(s.birth + s.span).toBeLessThanOrEqual(1.0001)
      for (const p of s.points) {
        expect(p.x).toBeGreaterThanOrEqual(-0.001)
        expect(p.x).toBeLessThanOrEqual(1.001)
        expect(p.y).toBeGreaterThanOrEqual(-0.001)
        expect(p.y).toBeLessThanOrEqual(1.001)
      }
    }
  })
})

function arcCenter(stroke: { points: Array<{ x: number; y: number }> }): { x: number; y: number } {
  const a = stroke.points[0]!
  const b = stroke.points[Math.floor(stroke.points.length / 2)]!
  const c = stroke.points[stroke.points.length - 1]!
  const d = 2 * (a.x * (b.y - c.y) + b.x * (c.y - a.y) + c.x * (a.y - b.y))
  const a2 = a.x * a.x + a.y * a.y
  const b2 = b.x * b.x + b.y * b.y
  const c2 = c.x * c.x + c.y * c.y
  return {
    x: (a2 * (b.y - c.y) + b2 * (c.y - a.y) + c2 * (a.y - b.y)) / d,
    y: (a2 * (c.x - b.x) + b2 * (a.x - c.x) + c2 * (b.x - a.x)) / d,
  }
}

function arcSweep(stroke: { points: Array<{ x: number; y: number }> }): number {
  const c = arcCenter(stroke)
  const a0 = Math.atan2(stroke.points[0]!.y - c.y, stroke.points[0]!.x - c.x)
  const a1 = Math.atan2(
    stroke.points[stroke.points.length - 1]!.y - c.y,
    stroke.points[stroke.points.length - 1]!.x - c.x,
  )
  let d = a1 - a0
  if (d < 0) d += 2 * Math.PI
  return d
}
